pkg/apis/pipeline/v1beta1/param_types.go

```diff
@@ -107,17 +132,23 @@ type ArrayOrString struct {
 	Type      ParamType `json:"type"` // Represents the stored type of ArrayOrString.
 	StringVal string    `json:"stringVal"`
 	// +listType=atomic
-	ArrayVal []string `json:"arrayVal"`
+	ArrayVal  []string          `json:"arrayVal"`
+	ObjectVal map[string]string `json:"objectVal"`
 }
 
 // UnmarshalJSON implements the json.Unmarshaller interface.
 func (arrayOrString *ArrayOrString) UnmarshalJSON(value []byte) error {
-	if value[0] == '"' {
+	switch value[0] {
+	case '"':
 		arrayOrString.Type = ParamTypeString
 		return json.Unmarshal(value, &arrayOrString.StringVal)
+	case '[':
+		arrayOrString.Type = ParamTypeArray
+		return json.Unmarshal(value, &arrayOrString.ArrayVal)
+	default:
+		arrayOrString.Type = ParamTypeObject
+		return json.Unmarshal(value, &arrayOrString.ObjectVal)
```

Comment:
Don't make this as default. Make a case for object and error at default case

Reply:
sgtm!

Reply:
I think we can treat all unsupported value to string. This is because the type and func is reused by results. For emitting results users may write integer to results path, in current code this will be read as string.

So to use ArrayOrString's UnmarshalJSON for emitting results and keep current behaviour not changed. I suggest we convert value to string if it doesn't fall to any switch condition or encountered err

I describe the case here:
https://github.com/tektoncd/pipeline/issues/4820

Reply:
Ah that makes sense. I think this thread can be closed.